### DISCUSSION TITLE
Digits only values recognised correctly, but sometimes stored wrong

### DISCUSSION THREAD

**Discussion by redman684** [jomjol/AI-on-the-edge-device#2993] [Q&A]
I'm reading my digital electricity meter. It's has only digits so I turned Anlalog ROI processing off.

![image](https://github.com/jomjol/AI-on-the-edge-device/assets/32646992/c94aa050-70e7-480d-af73-5f8f6752b5ce)

The transition from one number to the next is rather quick, the last digit quickly changes from 6 to 7 (in like a few seconds), then remains on 7 for a long time (generally about an hour) before it quickly changes to the next. It's rare a picture gets taken where the digits are not centered.

The problem can occur with any number, I've experienced it with 6-7, 7-8, 1-2 and 2-3 and it's always the higher number which is incorrectly determined to be the lower one.
It mostly works correctly with those numbers. I don't think it's OCR going wrong.

The digits get recognized correctly, but the resulting value is sometimes wrong.
![image](https://github.com/jomjol/AI-on-the-edge-device/assets/32646992/f7c0575e-0fda-49fd-882d-53627b89f481)

I can have days where it all goes perfectly, but sometimes it fluctuates wildly.

This also makes the graphs look ugly:
![image](https://github.com/jomjol/AI-on-the-edge-device/assets/32646992/c6e25151-b81b-44d8-874f-dc3ba9653d9a)

It this moment it seems unable to decide between 67626 and 67627.
The meter has been at 67627 for the entire hour and hasn't even been moving slightly, but the value stored fluctuates between 67626 and 67627.
![image](https://github.com/jomjol/AI-on-the-edge-device/assets/32646992/120e0300-e233-48cd-b135-1b7ddb970273)

In the data file I see the digit being read as 6.0,7.0,6.0,2.0,7.0, but it seems to get a +0,5 or -0.5 so the resulting value fluctuates ??

`2024-03-15T10:17:01+0100,main,67626,67626,67626,-0.500000,-1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:19:01+0100,main,67627,67627,67627,0.500000,1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:21:01+0100,main,67626,67626,67626,-0.500000,-1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:23:01+0100,main,67626,67626,67626,0.000000,0,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:25:01+0100,main,67626,67626,67626,0.000000,0,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:27:01+0100,main,67627,67627,67627,0.500000,1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:29:01+0100,main,67626,67626,67626,-0.500000,-1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:31:01+0100,main,67627,67627,67627,0.500000,1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:33:01+0100,main,67626,67626,67626,-0.500000,-1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:35:01+0100,main,67626,67626,67626,0.000000,0,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:37:01+0100,main,67627,67627,67627,0.500000,1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:39:01+0100,main,67626,67626,67626,-0.500000,-1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:41:01+0100,main,67626,67626,67626,0.000000,0,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:43:01+0100,main,67627,67627,67627,0.500000,1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:45:01+0100,main,67626,67626,67626,-0.500000,-1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:47:01+0100,main,67626,67626,67626,0.000000,0,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:49:01+0100,main,67626,67626,67626,0.000000,0,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:51:01+0100,main,67626,67626,67626,0.000000,0,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:53:01+0100,main,67626,67626,67626,0.000000,0,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:55:01+0100,main,67627,67627,67627,0.500000,1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:57:01+0100,main,67627,67627,67627,0.000000,0,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T10:59:01+0100,main,67626,67626,67626,-0.500000,-1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T11:01:01+0100,main,67627,67627,67627,0.500000,1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T11:03:01+0100,main,67627,67627,67627,0.000000,0,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T11:05:01+0100,main,67627,67627,67627,0.000000,0,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T11:07:01+0100,main,67626,67626,67626,-0.500000,-1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T11:09:01+0100,main,67626,67626,67626,0.000000,0,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T11:11:01+0100,main,67626,67626,67626,0.000000,0,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T11:13:01+0100,main,67627,67627,67627,0.500000,1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T11:15:01+0100,main,67627,67627,67627,0.000000,0,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T11:17:01+0100,main,67626,67626,67626,-0.500000,-1,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T11:19:01+0100,main,67626,67626,67626,0.000000,0,no error,6.0,7.0,6.0,2.0,7.0
2024-03-15T11:21:00+0100,main,67626,67626,67626,0.000000,0,no error,6.0,7.0,6.0,2.0,7.0`

What is going on here and how can I fix this.

It feels like tiny changes in image quality (shift?) might make the software recognize a 6.99999 or a 7.0000 and always round it down because further readings from analog dails are not available. But that's just me guessing.

I found this https://github.com/jomjol/AI-on-the-edge-device/discussions/2623 describing a similar problem, but no solution.

It did however trigger me to look at the Post-Processing settings 'Decimal Shift' (was set to 0) and 'Analog/Digital Transition Start' (was set to 9.2) and disabling both. 
The meter value has changed so I'll have to wait for the next time it occurs.a

**Comment by SybexX** [comment]
Try to improve the image quality, everything is so blurry

**Reply by redman684** [reply]
Sure, always a great tip, but it's clearly not misreading the digits. 

At the moment it flawlessly reads 6 7 6 2 9, but for some reason comes to a value of 67628

![image](https://github.com/jomjol/AI-on-the-edge-device/assets/32646992/d916c085-df81-44b4-b478-f5afb4bc0971)

![image](https://github.com/jomjol/AI-on-the-edge-device/assets/32646992/132d2fc5-016e-47fa-847d-e7fc441846e4)

**Reply by Slider0007** [reply]
@redman684 
I assume it's only a rounding issue of recognition details page. On this page values are rounded, internally values are always truncated.

Either you could enable extended resolution to get one more digit or move the ROI of last digit a little lower to ensure it's reading a value greater than zero and not hanging a little below zero crossing.

**Reply by redman684** [reply]
The Extended Resolution setting has "Use the decimal place of the last analog counter for increased accuracy." as description. As my meter doesn't have analog dials, this doesn't feel correct to do. 

The internal value seems to flutter around 8.996094 / 8.992188. but rounding either to 8 seems counter intuitive. 
I will try moving the ROIs down a bit and keep you updated.

**Reply by SybexX** [reply]
![Unbenannt](https://github.com/jomjol/AI-on-the-edge-device/assets/587201/fc47e00a-e0fb-4a8c-bdcb-992ad474078d)

**Reply by redman684** [reply]
My meter can run backwards, so I must allow negative values. 

Moving the POIs made no difference. Which seems kinda weird. (yes, I rebooted)
Using a high resolution class seems to allow for detecting fractions for digits which are higher/lower. So I'm indeed switching to a dig-class 11 model.
Will keep you informed.

**Reply by redman684** [reply]
dig-class11-700 seems to work better, except it's worse at reading the values (OCR). And when it can't read a value ('N') it makes them up instead of just not log a measurement. And it even misread a 4 as a 5, that's something the dig-class100 has never done wrong, it always recognized the correct digit.

I'll just start using the the recognized digits instead of whatever flawed value it makes out of it.

**Reply by SybexX** [reply]
Just try dig-class100 + Extended Resolution
and it doesn't matter whether AllowNegativeValues ​​is used or not

**Reply by redman684** [reply]
dig-class100 + Extended Resolution has a similar issue, except less strong. While I only have 5 digits and they're not moving at all, measurements keeps going back and forth between 65654.0 and 65653.9

> 2024-03-19T07:40:31+0100,main,67653.9,67653.9,67653.9,0.000000,0.0,no error,6.0,6.9,6.0,5.0,3.9
> 2024-03-19T07:42:31+0100,main,67654.0,67654.0,67654.0,0.050000,0.1,no error,6.0,6.9,5.8,5.0,4.0
> 2024-03-19T07:44:31+0100,main,67653.9,67653.9,67653.9,-0.050000,-0.1,no error,6.0,6.9,5.8,5.0,3.9
> 2024-03-19T07:46:31+0100,main,67654.0,67654.0,67654.0,0.050000,0.1,no error,6.0,6.9,6.0,5.0,4.0
> 2024-03-19T07:48:31+0100,main,67654.0,67654.0,67654.0,0.000000,0.0,no error,6.0,6.9,6.0,4.9,4.0
> 2024-03-19T07:50:32+0100,main,67653.9,67653.9,67653.9,-0.049587,-0.1,no error,6.0,6.9,6.0,4.9,3.9
> 2024-03-19T07:52:32+0100,main,67654.0,67654.0,67654.0,0.050000,0.1,no error,6.0,6.9,6.0,5.0,4.0
> 2024-03-19T07:54:32+0100,main,67653.9,67653.9,67653.9,-0.050000,-0.1,no error,6.0,6.9,5.8,5.0,3.9
> 2024-03-19T07:56:32+0100,main,67653.9,67653.9,67653.9,0.000000,0.0,no error,6.0,6.9,6.0,4.9,3.9
> 2024-03-19T07:58:32+0100,main,67653.9,67653.9,67653.9,0.000000,0.0,no error,6.0,6.9,6.0,4.9,3.9
> 2024-03-19T08:00:32+0100,main,67654.0,67654.0,67654.0,0.050000,0.1,no error,6.0,6.9,6.0,5.0,4.0
> 2024-03-19T08:02:32+0100,main,67654.0,67654.0,67654.0,0.000000,0.0,no error,6.0,6.9,5.8,4.9,4.0
> 2024-03-19T08:04:32+0100,main,67654.0,67654.0,67654.0,0.000000,0.0,no error,6.0,6.9,5.8,5.0,4.0
> 2024-03-19T08:06:32+0100,main,67653.9,67653.9,67653.9,-0.050000,-0.1,no error,6.0,6.9,5.8,5.0,3.9
> 2024-03-19T08:08:32+0100,main,67654.0,67654.0,67654.0,0.050000,0.1,no error,6.0,6.9,5.8,5.0,4.0
> 2024-03-19T08:10:32+0100,main,67653.9,67653.9,67653.9,-0.050000,-0.1,no error,6.0,6.9,6.0,5.0,3.9
> 2024-03-19T08:12:32+0100,main,67654.0,67654.0,67654.0,0.050000,0.1,no error,6.0,6.9,6.0,5.0,4.0
> 2024-03-19T08:14:32+0100,main,67653.9,67653.9,67653.9,-0.050000,-0.1,no error,6.0,6.9,5.8,5.0,3.9
> 2024-03-19T08:16:32+0100,main,67654.0,67654.0,67654.0,0.050000,0.1,no error,6.0,6.9,6.0,4.9,4.0
> 2024-03-19T08:18:32+0100,main,67654.0,67654.0,67654.0,0.000000,0.0,no error,6.0,6.9,5.8,4.9,4.0

Maybe I should put a sticker on the meter with a '0'.

I'll see if I can improve the image quality, but repeatedly removing the camera, adjusting the focus, placing it back and checking the image, is not an easy process.

**Reply by SybexX** [reply]
If you want, you can try the following firmware (backup your config.ini first)
[AI-on-the-edge-device__update__cam_settings2_min_(67a58e5).zip](https://github.com/jomjol/AI-on-the-edge-device/files/14650201/AI-on-the-edge-device__update__cam_settings2_min_.67a58e5.zip)

**Comment by SybexX** [comment]
But you don't really give the important information, such as your settings (which model do you use, prevalue on or off....)
and possibly enable debug to get more information in the log

**Reply by redman684** [reply]
The relevant portion of the debug file (I had to enable debug logging and wait for the issue to occur) seems to be:

Here it correctly reads a 9
> <DBG>	[CNN] ROI #4 - TfLite
> <DBG>	[CNN] CNN Type: DoubleHyprid10
> <DBG>	[CNN] After Invoke
> <DBG>	[CNN] _num (p, m): 9 0 8 _val (p, m): 0.996094 0.000000 0.003906 result: 8.996094 _fit: 1.000000
> <DBG>	[PSRAM] Shared memory used for Tensor Arena and model (PSRAM, part of shared memory) is free again
> <DBG>	[FLOWCTRL] Status: Post-Processing (17:36:17)
> <DBG>	[CNN] getReadout _analog=0, _extendedResolution=0, prev=-1
> <DBG>	[CNN] PointerEvalHybridNew - No predecessor - Result = 9 number: 8.996094 number_of_predecessors = -1.000000 eval_predecessors = -1 Digital_Uncertainty = 0.200000
> <DBG>	[CNN] getReadout(dig100)  prev=9

here it reads an 8
> <DBG>	[CNN] ROI #4 - TfLite
> <DBG>	[CNN] CNN Type: DoubleHyprid10
> <DBG>	[CNN] After Invoke
> <DBG>	[CNN] _num (p, m): 9 0 8 _val (p, m): 0.992188 0.000000 0.007812 result: 8.992188 _fit: 1.000000
> <DBG>	[PSRAM] Shared memory used for Tensor Arena and model (PSRAM, part of shared memory) is free again
> <DBG>	[FLOWCTRL] Status: Post-Processing (17:38:17)
> <DBG>	[CNN] getReadout _analog=0, _extendedResolution=0, prev=-1
> <DBG>	[CNN] PointerEvalHybridNew - No predecessor - Result = 8 number: 8.992188 number_of_predecessors = -1.000000 eval_predecessors = -1 Digital_Uncertainty = 0.200000
> <DBG>	[CNN] getReadout(dig100)  prev=8

does that mean it interperts the number as 8.996094 or 8.992188 (depending on the position?) one is rounded up while the other i rounded down? Seems rather arbitrary.

As for your other questions: I'm using a esp32cam with the Firmware version v15.7.0. prevalue is on.
